src/CookieMiddleware.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Cookies;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Psr\Log\LoggerInterface;
+use RuntimeException;
+use Yiisoft\Strings\WildcardPattern;
+
+use function explode;
+use function is_string;
+
+/**
+ * Validates and decrypts the values of the request cookie parameters,
+ * encrypts and signs the values of the response `Set-Cookie` headers.
+ */
+final class CookieMiddleware implements MiddlewareInterface
+{
+    private CookieEncryptor $encryptor;
+    private CookieSigner $signer;
+    private LoggerInterface $logger;
+
+    /**
+     * @var string[] The name patterns of the cookies that need to be encrypted/decrypted.
+     */
+    private array $encryption = [];
+
+    /**
+     * @var string[] The name patterns of the cookies that need to be signed/validated.
+     */
+    private array $signature = [];
+
+    /**
+     * @param LoggerInterface $logger The logger instance.
+     * @param CookieEncryptor $encryptor The encryptor instance.
+     * @param CookieSigner $signer The signer instance.
+     * @param string[] $namePatterns The array keys are cookie name patterns {@see \Yiisoft\Strings\WildcardPattern},
+     * and values are class names of {@see \Yiisoft\Cookies\CookieEncryptor} or {@see \Yiisoft\Cookies\CookieSigner}.
+     */
+    public function __construct(
+        LoggerInterface $logger,
+        CookieEncryptor $encryptor,
+        CookieSigner $signer,
+        array $namePatterns = []
```

Comment:
Parameters going first would allow passing namePatterns as positional constructor argument when configuring via Yii DI:

```suggestion
        array $namePatterns = [],
        LoggerInterface $logger,
        CookieEncryptor $encryptor,
        CookieSigner $signer
```

Also, `namePatterns` is better to be renamed to something like `cookiesSettings`.

Reply:
We can't put an optional parameter before the required ones, can we?

Reply:
You're right, we can't.